### ISSUE TITLE
Database migration

### ISSUE THREAD

**Issue by Gillani0** [nitrite/nitrite-java#213]
Is there any available tool for database migration, versioning etc that can be used with Nitrite?

Cheers

Syed

**Comment by anidotnet** [comment]
Can you clarify what kind of migration are you looking for? It is an embedded database and there is no breaking api changes till now. So database file created in v1.x can be used in v3.x library seamlessly.

**Comment by Gillani0** [comment]
Hi Anidotnet,

What I'm looking for is a tool similar to [Flybase](https://flywaydb.org/) or [Mangoc](https://github.com/cloudyrock/mongock) for version controlling or migrating data (if there are changes in classes etc).
Cheers

**Comment by anidotnet** [comment]
Unfortunately we don't have any such tools at this moment. 

You can use `NitriteCollection` where you can insert documents and you can change your dataset without worrying anything. If you are using `ObjectRepository` you can create it for an abstract class and you can change child class without breaking anything, provided you have not renamed any property of the class. You can also implement `Mappable` to change the mapping of your class dynamically.

**Comment by thegarlynch** [comment]
I think there is a need to reopen this, because implementation of serializable object might change. An idea is to make some option available for developer. for example :

- A destructive migration once version number change
- Destruction of document or repository once there is a conflict (this need new event for resynchronization, this is particularly needed in Object Repository)
- or a whole new mechanism other than document or repository that can add migration step

**Comment by Gillani0** [comment]
Hi,
I completely agree with this. Having a  dedicated migration tool is necessary since this will streamline and standardised the migration process. Cheers
Syed

**Comment by anidotnet** [comment]
OK, I'll take a look into this. Meantime, could you guys list out your use cases? Like - renaming class, or renaming fields etc.

And also what would be your preferred solution for migration? Something similar to android's room database migration where you will pass some custom migration rule in nitrite builder or something else?

**Comment by thegarlynch** [comment]
### Destructive Migration 

1. **General Destruction**, it is needed for application scope storage, it is especially useful for android
2. **With Version Number**, because Object Repository does not explicitly control schema of it's object class this is need either version number for all nitrite database or for particular object only. This might need a runtime check of data integrity (can be disabled)
3. **Collection or Object Repository destruction on conflict**, This need a new event, i propose that we should separate this with register callback and give it an option to call destructive only once (recommended) or multiple. We need to call this one time, because the application could be in always-conflicting-loop. But if we are going for version numbering route, the api kinda tell itself, and this doesn't really needed (at least for object repository)

### Proper Migration

#### Option 1 (Object Repository migration coupled to the object itself)
Since this is NoSQL, you doesn't need to make general migration like room or realm because repository changes is entirely coupled with object changes. The bad thing about this idea is : 
- Change could be transitive, the object itself might not change, but the class it referenced to might change
- It is hard to define good api for it. The only case it would work is for deleting column or make column with default implementation (just like any SQL migration). But transitive change might be harder to implement

An idea is to also made migration to transitive class that is affected. 
For example : 
```kotlin

@VersionNumber(3)
data class A(val column1 : String, val column2 : B)

class MigrationManagerForA {

   // i am copying realm here
  fun migrate(schema : Schema, fromVersion : Int, toVersion : Int){

    var version = fromVersion
    if(version == 1){
       schema.addWithDefault("column2", B(0, 0))
       // if B is changed, this code changes too
       // optionally : 
       //    schema.addWithDefaultJson("column", mapOf("field1" to 0, "field2" to 0))
       version = 2
    }
    if(version == 2){
       schema.doTransitiveMigration("column3", MigrationManagerForC(), fromVersion = 1, toVersion = 3)
       // another option is to just automatically use C migration manager
    }

}

data class B(val field1 : Int, val field2 : Int)

@VersionNumber(3)
data class C(val field1 : Long, val field2 : Int)

class MigrationManagerForC {

  fun migrate(document : Schema, fromVersion : Int, toVersion : Int){
      var version = fromVersion
      if(version == 1){
         schema.addWithDefault("field2", 0)
      }else {
         schema.raiseToLongInt("field1")
      }
  }

}

```

#### Option 2 (Entire new mechanism with fixed schema and only for type that is allowed just like SQL)

**Comment by Gillani0** [comment]
Hi,
I agree with the points raised by @thegarlynch.

From my PoV, I've worked and I really appreciate the way [mongobee](https://github.com/mongobee/mongobee) expose the DB migration APIs:
- Annotated methods with `ChangeLog` and `ChangeSet` with attributes to define the order or migration etc. 
- Extracting collection readable documents to update DB with updated class objects.

Since the concept of a general collection and a specific class-based Object repository already exists for Nitrate DB, the implementation should be simple: you'll need an embedded collection with a fixed schema that will track the versions and it's corresponding changes. 

Syed

**Comment by anidotnet** [comment]
Migration feature has been added in 4.x branch and it works as below:

```java

Migration migration = new Migration(Constants.INITIAL_SCHEMA_VERSION, 2) {
    @Override
    public void migrate(Instruction instruction) {
        instruction.forDatabase()
            .addPassword("test-user", "test-password");

        instruction.forRepository(OldClass.class, "demo1")
            .renameRepository("new", null)
            .changeDataType("empId", (TypeConverter<String, Long>) Long::parseLong)
            .changeIdField("uuid", "empId")
            .deleteField("uuid")
            .renameField("lastName", "familyName")
            .addField("fullName", document -> document.get("firstName", String.class) + " "
                + document.get("familyName", String.class))
            .dropIndex("firstName")
            .dropIndex("literature.text")
            .changeDataType("literature.ratings", (TypeConverter<Float, Integer>) Math::round);
    }
};

MVStoreModule storeModule = MVStoreModule.withConfig()
    .filePath(dbPath)
    .compress(true)
    .build();

db = Nitrite.builder()
    .loadModule(storeModule)
    .schemaVersion(2)
    .addMigrations(migration)
    .openOrCreate("test-user", "test-password");
```


The full demonstration can be found in one of the test codes here -  https://github.com/nitrite/nitrite-java/blob/4.x/nitrite-mvstore-adapter/src/test/java/org/dizitart/no2/migrate/MigrationTest.java

It is now available in 4.0.0-SNAPSHOT in maven snapshot repository.
